engine/src/main/java/org/terasology/engine/core/module/ModuleManager.java

```diff
@@ -105,49 +98,61 @@ public ModuleManager(String masterServerAddress, List<Class<?>> classesOnClasspa
 
     /**
      * I wondered why this is important, and found MovingBlocks/Terasology#1450.
-     * It's not a worry that the engine module wouldn't be loaded without it. 
-     * It's about ordering: some things run in an order derived from the dependency 
+     * It's not a worry that the engine module wouldn't be loaded without it.
+     * It's about ordering: some things run in an order derived from the dependency
      * tree, and we want to make sure engine is at the root of it.
-     */   
-    private void ensureModulesDependOnEngine(Module engineModule) {
+     */
+    private void ensureModulesDependOnEngine() {
         DependencyInfo engineDep = new DependencyInfo();
         engineDep.setId(engineModule.getId());
         engineDep.setMinVersion(engineModule.getVersion());
-        engineDep.setMaxVersion(engineModule.getVersion().getNextPatchVersion());
+        engineDep.setMaxVersion(engineModule.getVersion().getNextMinorVersion());
+
+        HashSet<Name> engineModules = Sets.newHashSet(engineModule.getId());
+        engineModules.addAll(engineModule.getMetadata().getDependencies().stream().map(DependencyInfo::getId).collect(Collectors.toList()));
```

Comment:
What are "engine modules", i.e., what are we collecting here?

Reply:
In _this specific context_ (the `ensureModulesDependOnEngine` method), this is–  

oh, I remember now, this was from the phase of this PR when we made a module for NUI, and so this `engineModules` here was "engine and _any modules engine depends on_" to avoid creating a circular dependency. e.g. engine depends on NUI, we don't want to add a dependency to NUI to make it depend on engine.

In this branch's current state, NUI is not a module, so we don't have to worry about that.

But also in this branch's current state, we have this [ridiculous list of hardcoded NUI packages and classes](https://github.com/MovingBlocks/Terasology/blob/1e9a36c7fcaabc513c9496bcbd32b888c1a163c5/engine/src/main/java/org/terasology/engine/core/module/ExternalApiWhitelist.java#L14), which according to some of the conversation around NUI 3.0 is a thing we should have some other way of doing: https://github.com/MovingBlocks/TeraNUI/pull/27#issuecomment-805639667

I remain very fuzzy on how to configure these "modules" that are libraries not subject to the same constraints we put on the stuff in `/modules/*`.

Reply:
I see - I was wondering whether these are subsystems or something like that. Maybe somebody needs to explain me the details of what we are doing, what the plan for maintaining it was/is, and what other (feasible) options we have... :see_no_evil: